src/audio/eq_fir/eq_fir_hifi3.c

```diff
@@ -33,55 +33,49 @@ void eq_fir_2x_s32(struct fir_state_32x16 fir[], struct input_stream_buffer *bso
 	ae_int32x2 d0 = 0;
 	ae_int32x2 d1 = 0;
 	ae_int32 *src = (ae_int32 *)source->r_ptr;
-	ae_int32 *snk = (ae_int32 *)sink->w_ptr;
-	ae_int32 *x;
-	ae_int32 *y0;
-	ae_int32 *y1;
+	ae_int32 *dst = (ae_int32 *)sink->w_ptr;
+	ae_int32 *in;
+	ae_int32 *out;
+	ae_int32 *out1;
 	int ch;
-	int i;
+	int i, n, nmax;
 	int rshift;
 	int lshift;
 	int shift;
 	int nch = source->channels;
 	int inc_nch_s = nch * sizeof(int32_t);
 	int inc_2nch_s = 2 * inc_nch_s;
-
-	for (ch = 0; ch < nch; ch++) {
-		/* Get FIR instance and get shifts.
-		 */
-		f = &fir[ch];
-		fir_get_lrshifts(f, &lshift, &rshift);
-		shift = lshift - rshift;
-
-		/* Copy src to x and advance src with dummy load */
-		fir_comp_setup_circular(source);
-		x = src;
-		AE_L32_XC(d0, src, sizeof(int32_t));
-
-		/* Copy snk to y0 and advance snk with dummy load. Pointer
-		 * y1 is set to be ahead of y0 with one frame.
-		 */
-		fir_comp_setup_circular(sink);
-		y0 = snk;
-		y1 = snk;
-		AE_L32_XC(d0, snk, sizeof(int32_t));
-		AE_L32_XC(d1, y1, inc_nch_s);
-
-		for (i = 0; i < (frames >> 1); i++) {
-			/* Load two input samples via input pointer x */
-			fir_comp_setup_circular(source);
-			AE_L32_XC(d0, x, inc_nch_s);
-			AE_L32_XC(d1, x, inc_nch_s);
-
-			/* Compute FIR */
+	int samples = nch * frames;
+
+	while (samples) {
```

Comment:
I'm worrying about change of processing order. A run in real device with long FIR would give more confidence that this is better. Maybe also xt testbench could be enough because it has pipelines and scheduler framework from SOF. You could add a test response where left and right channel have different coefficients and make them long. See the FIR EQ examples generator how it's done.